nextcord/client.py

```diff
@@ -2543,23 +2553,116 @@ def _add_decorated_application_commands(self) -> None:
+        .. versionchanged:: 2.3
+
+            This method is now documented.
+
+        Parameters
+        ----------
+        cog: :class:`nextcord.Cog`
+            The cog to register to the bot.
+        
+        override: :class:`bool`
+            If a previously loaded cog with the same name should be ejected
+            instead of raising an error.
+
+            .. versionadded:: 2.3
+
+        Raises
+        ------
+        TypeError
+            The cog does not inherit from :class:`Cog`.
+        CommandError
+            An error happened during loading.
+        ClientException
+            A cog with the same name is already loaded.
+        """
+
+        if not isinstance(cog, Cog):
+            raise TypeError("cogs must derive from nextcord.Cog")
+
+        cog_name = cog.__cog_name__
+        existing = self.__cogs.get(cog_name)
+
+        if existing is not None:
+            if not override:
+                raise ClientException(f"Cog named {cog_name!r} already loaded")
+            self.remove_cog(cog_name)
+
         for app_cmd in cog.application_commands:
             self.add_application_command(app_cmd, use_rollout=True)
 
-        self._client_cogs.add(cog)
+        self.__cogs[cog_name] = cog
 
-    def remove_cog(self, cog: Cog) -> None:
-        for app_cmd in cog.application_commands:
+    def get_cog(self, name: str) -> Optional[Cog]:
+        """Gets the cog instance requested.
+
+        If the cog is not found, ``None`` is returned instead.
+        
+        .. versionadded:: 2.3
+
+        Parameters
+        ----------
+        name: :class:`str`
+            The name of the cog you are requesting.
+            This is equivalent to the name passed via keyword
+            argument in class creation or the class name if unspecified.
+
+        Returns
+        -------
+        Optional[:class:`Cog`]
+            The cog that was requested. If not found, returns ``None``.
+        """
+        return self.__cogs.get(name)
+
+    def remove_cog(self, cog: Union[Cog, str]) -> Optional[Cog]:
+        """Removes a cog from the client and returns it.
+
+        All registered commands that the cog has registered 
+        will be removed as well.
+
+        If no cog is found then this method has no effect.
+
+        .. versionchanged:: 2.3
+
+            This method is now documented.
+
+        Parameters
+        ----------
+        name: Union[:class:`str`, :class:`Cog`]
+            Either the name of the cog to remove or the instance
+            of the cog to remove.
+
+        Returns
+        -------
+        Optional[:class:`nextcord.Cog`]
+             The cog that was removed. ``None`` if not found.
+        """
+        stored_cog: Optional[Cog] = None
+        if isinstance(cog, Cog):
+            for name, value in self.__cogs.items():
+                if value is cog:
+                    stored_cog = self.__cogs.pop(name)
+        else:
+            stored_cog = self.__cogs.pop(cog, None)
```

Comment:
Rather than iterating, can't it be done more efficiently in a single operation by checking for its unique cog name?
```suggestion
        if isinstance(cog, Cog):
            cog = cog.__cog_name__
        stored_cog = self.__cogs.pop(cog, None)
```

Reply:
How did I not think of that lmao